records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,187 @@
+
+[1] An individual commit in a database in not a problem usually, but when
+several of these occurs, potentially over 100,000 in the case of Notify in time
+of high traffic, it becomes a performance issue. An apt analogy is possible
+with someone getting to the cashier in a grocery store to pay their items:
+imagine that you would have to pay for each these items individually, i.e.
+a) scan one and only one item, b) ask for the total, c) pay with cash or card,
+d) repeat for all remaining items. This will take quite a long time compared
+to paying for all items in one batch within one transaction. Saving one
+notification every time in the database is akin to paying for one grocery item
+individually in one transaction, except you have potentially thousands of items
+remaining afterward. Ideally, we want to buy/save all for all of these in one
+go.
+
+[2] The EKS environment listens [on the `readinessProbe` health check](https://github.com/cds-snc/notification-manifests/blob/6ac406b315112d2b66d93d8fa74406c3e6e7639f/base/api-deployment.yaml#L216)
+that our setup exposes and it involves a direct database check. When the
+database slows down, so does the probe. The direct reaction from EKS is to
+detach the instances from the application's load balancers.
+
+_TODO: Add sequence or flow chart_
+
+## Options
+
+### Stream the notifications save
+
+Upon receiving a notification from the API, do not save it in the current
+*notification-api* process but defer it to a celery task. This does not
+address the root cause of our issue but offer many benefits, namely:
+
+1. This is an iterative improvement where we send the database save into
+   [Celery land](https://docs.celeryproject.org/en/stable/). Having these into
+   a Celery tasks means we can more easily control the flow of these operations
+   and control these with all of the Celery features. Specifically, we can batch
+   the notifications using [celery-batch](https://pypi.org/project/celery-batches/).
+2. The number of database connections will be lowered in the API component
+   as the saves moved into a Celery task.
+
+### Insert the notifications as a batch
+
+As we streamline the notification save with the previous option and got
+these under Celery land, we can now batch these easily. There is a
+[celery-batch](https://pypi.org/project/celery-batches/) project which
+exists that might offer us the option to easily achieve this with minimal
+setup.
+
+### Tweaking autovacuum options
+
+There are a few autovacuum configuration that we can tweak in order to delay
+the autovacuum action that triggers a performance slowdown of our database:
+
+* **autovacuum_scale_factor** (default: 0.2): we could use this to delay the vacuum, (suggestion 0.3).
+* **cost_delay**: We could use this to delay by increasing the threshold.
+
+### Manual execution of the vacuum process on the `notifications` table
+
+An option is to keep the autovacuum enabled but trigger a manual vacuum prior
```

Comment:
Should mention the con with this. We don't really know when we will get these bursts. We also don't know if the burst will still trigger the autovacc activity

Reply:
I can expand on this, mentioning that we need to identify the bursts. 

We do tend to know when these happen, as it's 95% of the time around 13h-15 EST. We cannot assume that will always be the case.

I personally do think that bursts will trigger autovacuum, i,e. the manual vacuums might relieve the autovacuum triggered by bursts, but I don't put much hope on it. We'd need to test and see, could be going against our expectations. 😀